packages/supi/src/api/storeAdd.ts

```diff
@@ -0,0 +1,59 @@
+import normalizeRegistryUrl = require('normalize-registry-url')
+
+import {streamParser} from '@pnpm/logger'
+import {StoreController} from 'package-store'
+
+import parseWantedDependencies from '../parseWantedDependencies';
+import {ReporterFunction} from '../types'
+
+export default async function (
+  fuzzyDeps: string[],
+  opts: {
+    prefix: string,
+    registry?: string,
+    verifyStoreIntegrity: boolean,
+    reporter?: ReporterFunction,
+    storeController: StoreController,
+  },
+) {
+  const reporter = opts && opts.reporter
+  if (reporter) {
+    streamParser.on('data', reporter)
+  }
+
+  const deps = await parseWantedDependencies(fuzzyDeps, {
+    allowNew: true,
+    currentPrefs: {},
+    defaultTag: '',
+    dev: false,
+    devDependencies: {},
+    optional: false,
+    optionalDependencies: {},
+  })
+
+  const pkgIds = []
+  for (const dep of deps) {
+    const ret = await opts.storeController.requestPackage(dep, {
```

Comment:
I think you can run them all simultaneously because that is how it works during regular installation (there is a network concurrency limit inside package-requester).

```typescript
const pkgIds = await Promise.all(deps.map(...
```